### PR TITLE
Ticket 31 : create form and post request

### PR DESCRIPTION
**Pull request by maximFeldman1**
i create form page using formik for adding product to our data
i create post request using msw,i send response to our server and get requ
Screen Shot 2022-12-24 at 22 12 18
est from server,then i adding our new product to our exist data.

### CODE REVIEW

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+`;
+
+const createProduct = (product: Partial<IProduct>) =>
+  ProductService.addProduct(product as IProduct);
+
+export const ProductForm = () => {
+  const { mutate } = useMutation(createProduct, {});
+
+  const handleFormSubmit = useCallback((data: Partial<IProduct>) => {
+    console.log(data, "dsd");
```
Comment (itairom-appsforce) [inline src/Components/UI/Forms/ProductForm/index.tsx]: remove console.log

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+  padding-top: 30px;
+`;
+
+const createProduct = (product: Partial<IProduct>) =>
+  ProductService.addProduct(product as IProduct);
+
+export const ProductForm = () => {
+  const { mutate } = useMutation(createProduct, {});
+
+  const handleFormSubmit = useCallback((data: Partial<IProduct>) => {
```
Comment (itairom-appsforce) [inline src/Components/UI/Forms/ProductForm/index.tsx]: rename to onSubmit

File: src/mockServer/controllers/products.ts [hunk 3]
```diff
@@ -12,4 +13,21 @@ export class ProductController {
...
     }
   }
+  static async addProduct(
+    req: RestRequest,
+    res: ResponseFunction,
+    ctx: RestContext
+  ) {
+    try {
+      const product = await req.json();
+      product.id = uuid();
```
Comment (itairom-appsforce) [inline src/mockServer/controllers/products.ts:23]:
```suggestion
product.id = req.id
```

File: src/Components/ProductsList/index.tsx [hunk 4]
```diff
@@ -1,10 +1,7 @@
 import { ProductService } from "../../shared/services/product-service";
```
Comment (amitmarko) [inline src/Components/ProductsList/index.tsx]: Please move `ProdcutsList` Component under Product page -> 
`pages/products/components/ProdcutsTable`

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 5]
```diff
@@ -0,0 +1,91 @@
+import React, { useCallback } from "react";
```
Comment (amitmarko) [inline src/Components/UI/Forms/ProductForm/index.tsx]: Please move `ProductForm` Component under Create Product page -> 
`pages/crearteProduct/components/ProductForm/index.tsx`


Let's talk Sunday morning on the folders structure.

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 2 above)
Comment (amitmarko) [reply src/Components/UI/Forms/ProductForm/index.tsx]: handleSubmit

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+const createProduct = (product: Partial<IProduct>) =>
+  ProductService.addProduct(product as IProduct);
+
+export const ProductForm = () => {
+  const { mutate } = useMutation(createProduct, {});
+
+  const handleFormSubmit = useCallback((data: Partial<IProduct>) => {
+    console.log(data, "dsd");
+    mutate(data);
```
Comment (amitmarko) [inline src/Components/UI/Forms/ProductForm/index.tsx]: Please create `ICreateProduct` interface (under product models) and prepare the Create Product 
```suggestion
    type ICreateProduct = Omit<Product, "id">;

   
    const newPolicy: ICreateProudct = {
       name: values.name,
       brand: values.brand,
       image: values.image,
       price:  values.price || 0
    }
    mutate(newPolicy);
```

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+
+  const formik = useFormik({
+    initialValues: {
+      name: " ",
+      brand: " ",
+      image: " ",
+    },
+
+    onSubmit: handleFormSubmit,
+  });
```
Comment (amitmarko) [inline src/Components/UI/Forms/ProductForm/index.tsx]: newLine

File: src/shared/index.tsx [hunk 8]
```diff
@@ -6,5 +6,6 @@ export const URL = {
 export const apiUrl = () => ({
   products: {
     getAll: "/api/products",
+    add: "/api/products/create",
```
Comment (amitmarko) [inline src/shared/index.tsx]: The api path is `create` therefore the key should be `create` as well.
```suggestion
    create: "/api/products/create",
```

File: src/shared/services/product-service.ts [hunk 9]
```diff
@@ -1,8 +1,12 @@
 import { apiUrl } from "..";
+import { IProduct } from "../../models";
 import Api from "./API";
 
 export class ProductService {
   static getAll() {
     return Api.get(apiUrl().products.getAll);
   }
+  static addProduct(product: IProduct) {
```
Comment (amitmarko) [inline src/shared/services/product-service.ts]:
```suggestion
  static createProduct(newProduct: ICreateProduct) {
```

File: src/Pages/CreateProduct/index.tsx [hunk 10]
```diff
@@ -1,5 +1,6 @@
 import React from "react";
+import { ProductForm } from "../../components/UI/Forms/ProductForm";
 
 export const CreateForm = () => {
```
Comment (amitmarko) [inline src/Pages/CreateProduct/index.tsx]: Why the name of the folder is `CreateProduct`  and the name of the component is `CreateForm`?
Please note all these things are super important
```suggestion
export const CreateProduct = () => {
```

File: src/Pages/CreateProduct/index.tsx [hunk 11]
```diff
@@ -1,5 +1,6 @@
 import React from "react";
+import { ProductForm } from "../../components/UI/Forms/ProductForm";
```
Comment (amitmarko) [inline src/Pages/CreateProduct/index.tsx]:
```suggestion
import { ProductForm } from "./components/ProductForm";
```

File: src/mockServer/routes/products.ts [hunk 12]
```diff
@@ -4,4 +4,5 @@ import { apiUrl } from "../../shared";
 
 export const productRoutes = [
   rest.get(apiUrl().products.getAll, ProductController.getAll),
+  rest.post(apiUrl().products.add, ProductController.addProduct),
```
Comment (amitmarko) [inline src/mockServer/routes/products.ts]:
```suggestion
  rest.post(apiUrl().products.create, ProductController.createProduct),
```

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+      <form onSubmit={formik.handleSubmit}>
+        <h1>Form Product</h1>
+        <span className="p-float-label mt-4">
+          <InputText
+            id="price"
+            name="price"
+            type="text"
+            onChange={formik.handleChange}
+            value={formik.values.price || " "}
+          />
```
Comment (amitmarko) [inline src/Components/UI/Forms/ProductForm/index.tsx]: Which fields are required? I think all..
Also Type should be number as `price` is number,
```suggestion
          <InputText
            id="price"
            name="price"
            type="number"
            required
            onChange={formik.handleChange}
            value={formik.values.price || ""}
          />
```

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 14]
```diff
@@ -0,0 +1,91 @@
...
+  justify-content: center;
+  align-items: center;
+  padding-top: 30px;
+`;
+
+const createProduct = (product: Partial<IProduct>) =>
+  ProductService.addProduct(product as IProduct);
+
+export const ProductForm = () => {
+  const { mutate } = useMutation(createProduct, {});
```
Comment (amitmarko) [inline src/Components/UI/Forms/ProductForm/index.tsx]:
```suggestion
  const { mutate } = useMutation(ProductService.createProduc, {});
```

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 15]
```diff
@@ -0,0 +1,91 @@
...
+
+const Root = styled.div`
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  padding-top: 30px;
+`;
+
+const createProduct = (product: Partial<IProduct>) =>
+  ProductService.addProduct(product as IProduct);
```
Comment (amitmarko) [inline src/Components/UI/Forms/ProductForm/index.tsx]:
```suggestion
```

File: src/shared/services/product-service.ts [hunk 16]
```diff
@@ -1,8 +1,12 @@
...
+    return Api.post<IProduct>(apiUrl().products.add, product);
```
Comment (amitmarko) [inline src/shared/services/product-service.ts]:
```suggestion
    return Api.post<IProduct>(apiUrl().products.create, newProduct);
```

File: src/models/product.ts [hunk 17]
```diff
@@ -0,0 +1,7 @@
+export interface IProduct {
+  id: string;
+  price: number;
+  name: string;
+  brand: string;
+  image: string;
+}
```
Comment (amitmarko) [inline src/models/product.ts:7]: Please create `ICreateProduct` as well,

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 18]
```diff
@@ -0,0 +1,91 @@
...
+        <span className="p-float-label mt-2">
+          <InputText
+            id="image"
+            name="image"
+            type="text"
+            onChange={formik.handleChange}
+            value={formik.values.image}
+          />
+          <label htmlFor="image">Image</label>
+        </span>
```
Comment (amitmarko) [inline src/Components/UI/Forms/ProductForm/index.tsx]: Please create out `Input` component on `components` folder with span + InputText + label,
do not repeat yourself

File: src/Components/UI/Forms/ProductForm/index.tsx [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+
+export const ProductForm = () => {
+  const { mutate } = useMutation(createProduct, {});
+
+  const handleFormSubmit = useCallback((data: Partial<IProduct>) => {
+    console.log(data, "dsd");
+    mutate(data);
+  }, []);
+
+  const formik = useFormik({
```
Comment (amitmarko) [inline src/Components/UI/Forms/ProductForm/index.tsx]: Please create `yup` schema with validation

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 1 above)
Comment (maximFeldman1) [reply src/Components/UI/Forms/ProductForm/index.tsx]: Fixed

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 2 above)
Comment (maximFeldman1) [reply src/Components/UI/Forms/ProductForm/index.tsx]: Fixed

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 7 above)
Comment (maximFeldman1) [reply src/Components/UI/Forms/ProductForm/index.tsx]: Fixed

File: src/shared/index.tsx (hunk 8 above)
Comment (maximFeldman1) [reply src/shared/index.tsx]: Fixed

File: src/shared/services/product-service.ts (hunk 9 above)
Comment (maximFeldman1) [reply src/shared/services/product-service.ts]: Fixed

File: src/Pages/CreateProduct/index.tsx (hunk 10 above)
Comment (maximFeldman1) [reply src/Pages/CreateProduct/index.tsx]: Fixed

File: src/mockServer/routes/products.ts (hunk 12 above)
Comment (maximFeldman1) [reply src/mockServer/routes/products.ts]: Fixed

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 13 above)
Comment (maximFeldman1) [reply src/Components/UI/Forms/ProductForm/index.tsx]: Fixed

File: src/shared/services/product-service.ts (hunk 16 above)
Comment (maximFeldman1) [reply src/shared/services/product-service.ts]: Fixed

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 18 above)
Comment (maximFeldman1) [reply src/Components/UI/Forms/ProductForm/index.tsx]: Fixed

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 19 above)
Comment (maximFeldman1) [reply src/Components/UI/Forms/ProductForm/index.tsx]: Fixed

File: src/models/product.ts (hunk 17 above)
Comment (maximFeldman1) [reply src/models/product.ts:7]: Fixed

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 15 above)
Comment (maximFeldman1) [reply src/Components/UI/Forms/ProductForm/index.tsx]: i leave it like that was, call createProduct inside the useMutation

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 14 above)
Comment (maximFeldman1) [reply src/Components/UI/Forms/ProductForm/index.tsx]: Fixed

File: src/Components/UI/Forms/ProductForm/index.tsx (hunk 5 above)
Comment (maximFeldman1) [reply src/Components/UI/Forms/ProductForm/index.tsx]: Fixed

File: src/Components/ProductsList/index.tsx (hunk 4 above)
Comment (maximFeldman1) [reply src/Components/ProductsList/index.tsx]: Fixed

File: src/Components/UI/Forms/InputField/index.tsx [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+
+export const InputField = ({
+  value,
+  type,
+  name,
+  errors,
+  ...defaultProps
+}: any) => {
+  return (
+    <>
```
Comment (amitmarko) [inline src/Components/UI/Forms/InputField/index.tsx:13]:
```suggestion
```

File: src/Components/UI/Forms/InputField/index.tsx [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+      <span className="p-float-label mt-4">
+        <InputText name={name} type={type} value={value} {...defaultProps} />
+        <label htmlFor={name}>{name}</label>
+        {<div>{errors}</div>}
+      </span>
+    </>
```
Comment (amitmarko) [inline src/Components/UI/Forms/InputField/index.tsx:19]:
```suggestion
```

File: src/Components/UI/Forms/InputField/index.tsx [hunk 22]
```diff
@@ -0,0 +1,21 @@
...
+  name,
+  errors,
+  ...defaultProps
+}: any) => {
+  return (
+    <>
+      <span className="p-float-label mt-4">
+        <InputText name={name} type={type} value={value} {...defaultProps} />
+        <label htmlFor={name}>{name}</label>
+        {<div>{errors}</div>}
```
Comment (amitmarko) [inline src/Components/UI/Forms/InputField/index.tsx:17]:
```suggestion
        <div>{errors}</div>
```

File: src/Components/UI/Forms/InputField/index.tsx [hunk 23]
```diff
@@ -0,0 +1,21 @@
...
+import { InputText } from "primereact/inputtext";
+import { ErrorMessage } from "formik";
+
+export const InputField = ({
+  value,
+  type,
+  name,
+  errors,
+  ...defaultProps
+}: any) => {
```
Comment (amitmarko) [inline src/Components/UI/Forms/InputField/index.tsx:11]: Sit with Shalev about interface based on Prime type

File: src/Pages/CreateProduct/components/ProductForm/index.tsx [hunk 24]
```diff
@@ -0,0 +1,89 @@
...
+
+const Root = styled.div`
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  padding-top: 30px;
+`;
+
+const createProduct = (product: Partial<IProduct>) =>
+  ProductService.createProduct(product as IProduct);
```
Comment (amitmarko) [inline src/Pages/CreateProduct/components/ProductForm/index.tsx]:
```suggestion
  ProductService.createProduct(newProduct);
```

File: src/Pages/CreateProduct/components/ProductForm/index.tsx [hunk 25]
```diff
@@ -0,0 +1,89 @@
...
+import { validationYup } from "../../../../schema/validations/validationSchema";
+
+const Root = styled.div`
+  display: flex;
+  justify-content: center;
+  align-items: center;
+  padding-top: 30px;
+`;
+
+const createProduct = (product: Partial<IProduct>) =>
```
Comment (amitmarko) [inline src/Pages/CreateProduct/components/ProductForm/index.tsx]:
```suggestion
const createProduct = (newProduct: ICreateProduct) =>
```

File: src/Pages/CreateProduct/components/ProductForm/index.tsx [hunk 26]
```diff
@@ -0,0 +1,89 @@
...
+
+  const formik = useFormik({
+    initialValues: {
+      name: " ",
+      brand: " ",
+      image: " ",
+    },
+    onSubmit: handleSubmit,
+    validationSchema: validationYup,
+  });
```
Comment (amitmarko) [inline src/Pages/CreateProduct/components/ProductForm/index.tsx:35]: new line

File: src/Pages/CreateProduct/components/ProductForm/index.tsx [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+        <InputField
+          name="brand"
+          type="text"
+          onChange={formik.handleChange}
+          onBlur={formik.handleBlur}
+          value={formik.values.brand}
+          errors={formik.errors.brand}
+          required
+        />
+
```
Comment (amitmarko) [inline src/Pages/CreateProduct/components/ProductForm/index.tsx]:
```suggestion
```

**Pull request merged**